Python/sysmodule.c

```diff
@@ -96,6 +96,80 @@ PySys_SetObject(const char *name, PyObject *v)
         return PyDict_SetItemString(sd, name, v);
 }
 
+static PyObject *
+sys_breakpointhook(PyObject *self, PyObject *args, PyObject *keywords)
+{
```

Comment:
Add "assert(!PyErr_Occurred());" at the entry point since this function is called when something goes wrong, and the implementation don't expect that an exception is currently set.

... Maybe even raise an exception at runtime if it's called with an exception set? See _Py_CheckFunctionResult() which implements a similar check at *exit* point.

Reply:
Is that true, i.e. that it's possible to call this with an exception set?  I don't mind adding an assertion for that, but I'm not sure I see how it's possible.  `builtin_breakpoint()` won't call it with an exception set AFAICT.  What scenario are you thinking about?

Reply:
> Is that true, i.e. that it's possible to call this with an exception set?

In my experience, it's very common that a function returning a result *and* set an exception is not catched immediately, but somewhere "later". I had such issues when I ran my pyfailmalloc tool which injects MemoryError anywhere.

But I made many changes in CPython internals to add assert(!PyErr_Occurred()) at many places. The main change is that all functions calling functions should now check that no exception is raised if the called function returned a result. This is _Py_CheckFunctionResult().

To come back to your breakpoint() implementation, "assert(!PyErr_Occurred());" should be enough.

Reply:
Just to be clear, you're suggesting `assert(!PyErr_Occurred());` at the beginning of `sys_breakpointhook()` right?  I'm cool with that.

Reply:
Yes.